File: mobile-hospital/src/features/growth/growthCharacteristics.ts
export type GrowthCharacteristics = {
  phrase: string;
  labels: string[];
  traitCodes: string[];
};

const ADJECTIVE_TRAITS = new Set(['LEAN', 'STOCKY', 'SLENDER', 'TALL', 'SHORT', 'LARGE_HEAD']);

function traitLabelKey(code: string): string {
  return `growth.characteristic.${code.toLowerCase()}`;
}

export function deriveGrowthTraitCodes(record: {
  weightPercentile?: number | null;
  heightPercentile?: number | null;
  bmiPercentile?: number | null;
  hcPercentile?: number | null;
}): string[] {
  const codes: string[] = [];
  const wp = record.weightPercentile;
  const hp = record.heightPercentile;
  const bp = record.bmiPercentile;
  const hcp = record.hcPercentile;

  if (bp != null && bp < 15 && (hp == null || hp >= 50)) codes.push('LEAN');
  else if (bp != null && bp > 85) codes.push('STOCKY');
  else if (wp != null && wp < 15) codes.push('SLENDER');

  if (hp != null && hp >= 85) codes.push('TALL');
  else if (hp != null && hp <= 15) codes.push('SHORT');

  if (wp != null) {
    if (wp >= 15 && wp <= 85) codes.push('NORMAL_WEIGHT');
    else if (wp < 15) codes.push('UNDERWEIGHT');
    else codes.push('OVERWEIGHT');
  }
  if (hcp != null && hcp >= 85) codes.push('LARGE_HEAD');
  if (!codes.some((code) => ADJECTIVE_TRAITS.has(code)) && !codes.length) codes.push('NORMAL_WEIGHT');

  return [...new Set(codes)];
}

function resolveSexLabel(sex: string | null | undefined, label: (key: string) => string): string {
  const sexNorm = String(sex ?? '').trim().toLowerCase();
  if (sexNorm === 'female') return label('growth.characteristic.girl');
  if (sexNorm === 'male') return label('growth.characteristic.boy');
  return label('growth.characteristic.child');
}

export function buildGrowthProfilePhrase(
  traitCodes: string[],
  sex: string | null | undefined,
  label: (key: string) => string
): string {
  const sexLabel = resolveSexLabel(sex, label);
  const adjectives = traitCodes
    .filter((code) => ADJECTIVE_TRAITS.has(code))
    .map((code) => label(traitLabelKey(code)).toLowerCase());

  if (adjectives.length) {
    let phrase = adjectives[0].charAt(0).toUpperCase() + adjectives[0].slice(1);
    if (adjectives.length > 1) phrase += ` ${adjectives.slice(1).join(' ')}`;
    return `${phrase} ${sexLabel.toLowerCase()}`.trim();
  }
  return sexLabel.trim();
}

function buildGrowthLabels(
  traitCodes: string[],
  sex: string | null | undefined,
  label: (key: string) => string
): string[] {
  const labels = traitCodes.map((code) => label(traitLabelKey(code)));
  const sexLabel = resolveSexLabel(sex, label);
  if (sexLabel && !labels.includes(sexLabel)) labels.push(sexLabel);
  return labels;
}

export function deriveGrowthCharacteristics(
  sex: string | null | undefined,
  record: {
    weightPercentile?: number | null;
    heightPercentile?: number | null;
    bmiPercentile?: number | null;
    hcPercentile?: number | null;
  },
  label: (key: string) => string
): GrowthCharacteristics {
  const traitCodes = deriveGrowthTraitCodes(record);
  return {
    phrase: buildGrowthProfilePhrase(traitCodes, sex, label),
    labels: buildGrowthLabels(traitCodes, sex, label),
    traitCodes
  };
}

export function resolveGrowthCharacteristics(
  sex: string | null | undefined,
  record: {
    weightPercentile?: number | null;
    heightPercentile?: number | null;
    bmiPercentile?: number | null;
    hcPercentile?: number | null;
  },
  label: (key: string) => string,
  fromApi?: GrowthCharacteristics | null
): GrowthCharacteristics {
  if (fromApi?.traitCodes?.length) {
    return {
      traitCodes: fromApi.traitCodes,
      labels: buildGrowthLabels(fromApi.traitCodes, sex, label),
      phrase: buildGrowthProfilePhrase(fromApi.traitCodes, sex, label)
    };
  }
  if (fromApi?.phrase) return fromApi;
  return deriveGrowthCharacteristics(sex, record, label);
}

export function parseGrowthCharacteristics(raw: unknown): GrowthCharacteristics | null {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return null;
  const row = raw as Record<string, unknown>;
  const phrase = String(row.Phrase ?? row.phrase ?? '').trim();
  const labelsRaw = row.Labels ?? row.labels;
  const codesRaw = row.TraitCodes ?? row.traitCodes;
  const labels = Array.isArray(labelsRaw) ? labelsRaw.map((v) => String(v)) : [];
  const traitCodes = Array.isArray(codesRaw) ? codesRaw.map((v) => String(v)) : [];
  if (!phrase && !labels.length && !traitCodes.length) return null;
  return { phrase, labels, traitCodes };
}

File: pdf-rag-pipeline/query/growth_characteristics.py
from __future__ import annotations

from typing import Any, Optional

ADJECTIVE_TRAITS = frozenset({"LEAN", "STOCKY", "SLENDER", "TALL", "SHORT", "LARGE_HEAD"})

_TRAIT_LABELS: dict[str, dict[str, str]] = {
    "en": {
        "LEAN": "Lean",
        "STOCKY": "Stocky",
        "SLENDER": "Slender",
        "TALL": "Tall",
        "SHORT": "Short",
        "NORMAL_WEIGHT": "Normal weight",
        "UNDERWEIGHT": "Underweight",
        "OVERWEIGHT": "Overweight",
        "LARGE_HEAD": "Large head",
        "BOY": "Boy",
        "GIRL": "Girl",
        "CHILD": "Child",
    },
    "hi": {
        "LEAN": "दुबला",
        "STOCKY": "मोटा",
        "SLENDER": "पतला",
        "TALL": "लंबा",
        "SHORT": "छोटा",
        "NORMAL_WEIGHT": "सामान्य वजन",
        "UNDERWEIGHT": "कम वजन",
        "OVERWEIGHT": "अधिक वजन",
        "LARGE_HEAD": "बड़ा सिर",
        "BOY": "लड़का",
        "GIRL": "लड़की",
        "CHILD": "बच्चा",
    },
    "kn": {
        "LEAN": "ತೆಳ್ಳಗಿನ",
        "STOCKY": "ದಪ್ಪ",
        "SLENDER": "ಸಣಕ",
        "TALL": "ಎತ್ತರ",
        "SHORT": "ಕುಳ್ಳ",
        "NORMAL_WEIGHT": "ಸಾಮಾನ್ಯ ತೂಕ",
        "UNDERWEIGHT": "ಕಡಿಮೆ ತೂಕ",
        "OVERWEIGHT": "ಹೆಚ್ಚು ತೂಕ",
        "LARGE_HEAD": "ದೊಡ್ಡ ತಲೆ",
        "BOY": "ಹುಡುಗ",
        "GIRL": "ಹುಡುಗಿ",
        "CHILD": "ಮಗು",
    },
}


def _label(locale: str, code: str) -> str:
    bucket = _TRAIT_LABELS.get(locale) or _TRAIT_LABELS["en"]
    return bucket.get(code) or _TRAIT_LABELS["en"].get(code, code)


def derive_trait_codes(
    *,
    weight_percentile: Optional[float],
    height_percentile: Optional[float],
    bmi_percentile: Optional[float],
    hc_percentile: Optional[float],
) -> list[str]:
    codes: list[str] = []
    wp = float(weight_percentile) if weight_percentile is not None else None
    hp = float(height_percentile) if height_percentile is not None else None
    bp = float(bmi_percentile) if bmi_percentile is not None else None
    hcp = float(hc_percentile) if hc_percentile is not None else None

    if bp is not None and bp < 15.0 and (hp is None or hp >= 50.0):
        codes.append("LEAN")
    elif bp is not None and bp > 85.0:
        codes.append("STOCKY")
    elif wp is not None and wp < 15.0:
        codes.append("SLENDER")

    if hp is not None and hp >= 85.0:
        codes.append("TALL")
    elif hp is not None and hp <= 15.0:
        codes.append("SHORT")

    if wp is not None:
        if 15.0 <= wp <= 85.0:
            codes.append("NORMAL_WEIGHT")
        elif wp < 15.0:
            codes.append("UNDERWEIGHT")
        else:
            codes.append("OVERWEIGHT")

    if hcp is not None and hcp >= 85.0:
        codes.append("LARGE_HEAD")

    if not any(code in ADJECTIVE_TRAITS for code in codes) and not codes:
        codes.append("NORMAL_WEIGHT")

    # Preserve order, drop duplicates
    seen: set[str] = set()
    ordered: list[str] = []
    for code in codes:
        if code not in seen:
            seen.add(code)
            ordered.append(code)
    return ordered


def derive_growth_characteristics(
    *,
    sex: Optional[str],
    weight_percentile: Optional[float],
    height_percentile: Optional[float],
    bmi_percentile: Optional[float],
    hc_percentile: Optional[float],
    reply_locale: Optional[str] = None,
) -> dict[str, Any]:
    locale = str(reply_locale or "en").strip().lower()[:2] or "en"
    trait_codes = derive_trait_codes(
        weight_percentile=weight_percentile,
        height_percentile=height_percentile,
        bmi_percentile=bmi_percentile,
        hc_percentile=hc_percentile,
    )
    labels = [_label(locale, code) for code in trait_codes]

    sex_norm = str(sex or "").strip().lower()
    if sex_norm == "female":
        sex_label = _label(locale, "GIRL")
    elif sex_norm == "male":
        sex_label = _label(locale, "BOY")
    else:
        sex_label = _label(locale, "CHILD")

    if sex_label and sex_label not in labels:
        labels.append(sex_label)

    adjectives = [
        _label(locale, code).lower()
        for code in trait_codes
        if code in ADJECTIVE_TRAITS
    ]
    if adjectives:
        phrase = adjectives[0].capitalize()
        if len(adjectives) > 1:
            phrase += " " + " ".join(adjectives[1:])
        phrase += f" {sex_label.lower()}"
    else:
        phrase = sex_label

    return {
        "Phrase": phrase.strip(),
        "Labels": labels,
        "TraitCodes": trait_codes,
    }
